src/pid.cpp

```diff
@@ -408,24 +285,67 @@ void Pid::getCurrentPIDErrors(double *pe, double *ie, double *de)
   *de = d_error_;
 }
 
-void Pid::printValues()
+void Pid::printValues(const rclcpp::Logger & logger)
 {
   Gains gains = getGains();
 
-  ROS_INFO_STREAM_NAMED("pid","Current Values of PID Class:\n"
-    << "  P Gain: " << gains.p_gain_ << "\n"
-    << "  I Gain: " << gains.i_gain_ << "\n"
-    << "  D Gain: " << gains.d_gain_ << "\n"
-    << "  I_Max:  " << gains.i_max_  << "\n"
-    << "  I_Min:  " << gains.i_min_  << "\n"
-    << "  Antiwindup:  " << gains.antiwindup_  << "\n"
-    << "  P_Error_Last: " << p_error_last_  << "\n"
-    << "  P_Error:      " << p_error_  << "\n"
-    << "  I_Error:       " << i_error_  << "\n"
-    << "  D_Error:      " << d_error_  << "\n"
-    << "  Command:      " << cmd_
+  RCLCPP_INFO_STREAM(logger,
+    "Current Values of PID Class:\n" <<
+    "  P Gain:       " << gains.p_gain_ << "\n" <<
+    "  I Gain:       " << gains.i_gain_ << "\n" <<
+    "  D Gain:       " << gains.d_gain_ << "\n" <<
+    "  I_Max:        " << gains.i_max_ << "\n" <<
+    "  I_Min:        " << gains.i_min_ << "\n" <<
+    "  Antiwindup:   " << gains.antiwindup_ << "\n" <<
+    "  P_Error_Last: " << p_error_last_ << "\n" <<
+    "  P_Error:      " << p_error_ << "\n" <<
+    "  I_Error:      " << i_error_ << "\n" <<
+    "  D_Error:      " << d_error_ << "\n" <<
+    "  Command:      " << cmd_
   );
+}
 
+void Pid::setParameterEventCallback()
+{
+  auto on_parameter_event_callback = [this](const std::vector<rclcpp::Parameter> & parameters) {
+      rcl_interfaces::msg::SetParametersResult result;
+      result.successful = true;
+
+      /// @note don't use getGains, it's rt
+      Gains gains = *gains_buffer_.readFromNonRT();
+
+      for (auto & parameter : parameters) {
+        const std::string param_name = parameter.get_name();
+        if (param_name == "p") {
+          gains.p_gain_ = parameter.get_value<double>();
+        } else if (param_name == "i") {
+          gains.i_gain_ = parameter.get_value<double>();
+        } else if (param_name == "d") {
+          gains.d_gain_ = parameter.get_value<double>();
+        } else if (param_name == "i_clamp_max") {
```

Comment:
There was a parameter i_clamp from melodic_devel, do you mind putting it in?

https://github.com/ros-controls/control_toolbox/blob/melodic-devel/src/pid.cpp#L126

Reply:
Do we really need this? It seems kind of redundant.

Reply:
well someone who had the i_clamp parameter and was using it in ros1 would need to substitute it for 2 more parameters when porting to ros2 - and tracking it down would not be very obvious especially when we don't generally have the time to put up changelists. Also I think we can spare the few lines of code.

Reply:
I'd argue that a big change such as switching to ROS2 would be the perfect time to address this type of inconsistencies since the user will need to adapt a lot of code anyways.

Would you agree to discuss this as part of #85? I think this PR has been delayed enough already.

Reply:
Edit: just saw that @bmagyar has already merged.